Sources/App/Auth/Models/AppUser.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  File.swift
+//  
+//
+//  Created by Joe Williams on 14/11/2021.
+//
+
+import Foundation
+import Vapor
+import Fluent
+
+final class AppUser: Model, ModelAuthenticatable, Content {
+    
+    enum UserRole: String, Codable {
+        case user, editor, admin
+    }
+    
+    static let usernameKey = \AppUser.$email
+    static let passwordHashKey = \AppUser.$passwordHash
+    static let schema = "app_users"
+    
+    // Unique identifier for this user.
+    @ID(key: .id)
+    var id: UUID?
+
+    // The user's name.
+    @Field(key: "name")
+    var name: String
+    
+    @Field(key: "email")
+    var email: String
+    
+    @Field(key: "password_hash")
+    var passwordHash: String
+    
+    @Field(key: "user_role")
+    var role: UserRole
+    
+    // Creates a new, empty user.
+    init() { }
+
+    // Creates a new user with all properties set.
+    init(id: UUID? = nil, name: String, email: String, passwordHash: String, role: UserRole) {
+        self.id = id
+        self.name = name
+        self.email = email
+        self.passwordHash = passwordHash
+        self.role = role
+    }
+    
+    func verify(password: String) throws -> Bool {
+        try Bcrypt.verify(password, created: self.passwordHash)
+    }
+    
+    func generateToken() throws -> UserToken {
+        try .init(
+            value: [UInt8].random(count: 16).base64,
+            timestamp: Date().addingTimeInterval(604800), // token is valid for 1 week
+            userID: self.requireID()
+        )
+    }
+    
+    class Migrations: Migration {
```

Comment:
Might be good to keep migrations separate, then makes it easier as new migrations are needed in the future.

Reply:
how does that look @pdamonkey? i've shamelessly copied the examples used in vapor docs

Reply:
I'd expect you'd have a Migrations group and then either have AppUserMigrations file in there, or as per the book, CreateAppUser a a migration, then perhaps other ones to update after that?